### PR TITLE
Add Model and training

### PR DESCRIPTION
**Pull request by shahbuland**
Overview of additions:
- Add configs for model and training
- Add example config and example training script for story critique dataset
- Add  base cross encoder class
- Add text-text cross encoder
- Add generic trainer  for any cross encoder model (not accelerate or deepspeed *yet*)

- Add img-text cross encoder
- Add RPCD dataset pipeline

Changes:
- Pipeline preprocessing needs to generalize to cases where preprocessing happens before or after tokenization/feature extraction. To solve, just use one preprocess function rather than previous method of having two. This is simpler in code anyways.
- Pipeline now has validation loader and validation set partitioning  
  
TODO:
- Deepspeed support
- Proper WANDB testing
- Checkpoint testing
- Reranking testing
- Modality encodings

### CODE REVIEW

**Comment by shahbuland** [comment]
Merging to create initial framework for Magicarp v2. Further PRs will require review and main will become a protected branch.
